swap/swap.go

```diff
@@ -17,69 +17,340 @@
+	if err != nil {
+		log.Error("error while creating cheque: %s", err.Error())
+		return err
+	}
+
+	log.Info(fmt.Sprintf("sending cheque with serial %d, amount %d, benficiary %v, contract %v", cheque.ChequeParams.Serial, cheque.ChequeParams.Amount, cheque.Beneficiary, cheque.Contract))
+	s.cheques[peer] = cheque
+
+	err = s.stateStore.Put(peer.String()+"_cheques", &cheque)
+	// TODO: error handling might be quite more complex
+	if err != nil {
+		log.Error("error while storing the last cheque: %s", err.Error())
+		return err
+	}
+
+	emit := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+
+	// reset balance;
+	// TODO: if sending fails it should actually be roll backed...
+	s.resetBalance(peer, int64(cheque.Amount))
+
+	err = swapPeer.Send(context.TODO(), emit)
 	return err
 }
 
-//GetPeerBalance returns the balance for a given peer
-func (swap *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
-	swap.lock.RLock()
-	defer swap.lock.RUnlock()
-	if p, ok := swap.balances[peer]; ok {
-		return p, nil
+// Create a Cheque structure emitted to a specific peer as a beneficiary
+// The serial and amount of the cheque will depend on the last cheque and current balance for this peer
+// The cheque will be signed and point to the issuer's contract
+func (s *Swap) createCheque(peer enode.ID) (*Cheque, error) {
+	var cheque *Cheque
+	var err error
+
+	swapPeer := s.getPeer(peer)
+	beneficiary := swapPeer.beneficiary
+
+	peerBalance := s.balances[peer]
+	// the balance should be negative here, we take the absolute value:
+	honey := uint64(-peerBalance)
+
+	// convert honey to ETH
+	var amount uint64
+	amount, err = s.oracle.GetPrice(honey)
+	if err != nil {
+		log.Error("error getting price from oracle", "err", err)
+		return nil, err
+	}
+
+	// we need to ignore the error check when loading from the StateStore,
+	// as an error might indicate that there is no existing cheque, which
+	// could mean it's the first interaction, which is absolutely valid
+	_ = s.loadCheque(peer)
+	lastCheque := s.cheques[peer]
+
+	if lastCheque == nil {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: uint64(1),
+				Amount: uint64(amount),
+			},
+		}
+	} else {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: lastCheque.Serial + 1,
+				Amount: lastCheque.Amount + uint64(amount),
+			},
+		}
+	}
+	cheque.ChequeParams.Timeout = defaultCashInDelay
+	cheque.ChequeParams.Contract = s.owner.Contract
+	cheque.ChequeParams.Honey = uint64(honey)
+	cheque.Beneficiary = beneficiary
+
+	cheque.Sig, err = s.signContent(cheque)
+
+	return cheque, err
+}
+
+// PeerBalance returns the balance for a given peer
+func (s *Swap) PeerBalance(peer enode.ID) (int64, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	var peerBalance int64
+	var err error
+
+	// load balance from memory
+	peerBalance, keyExists := s.balances[peer]
+	if !keyExists {
+		// if not in memory, try to load balance from disk
+		err = s.stateStore.Get(peer.String(), &peerBalance)
```

Comment:
check the error. i'm not a big fan of having the `err` variable defined in the function signature then set it around the function body. it makes flow control very difficult. better deal with the errors when they arise as they might require different treatment (so try to return errors as fast as possible rather than letting the function run till the end)

Reply:
Fixed.